docs/programming-guide.md

```diff
@@ -1086,6 +1086,62 @@ for details.
 </tr>
 </table>
 
+### Shuffle operations
+
+Certain operations within Spark trigger an event known as the shuffle. The shuffle is Spark's 
+mechanism for re-distributing data so that is grouped differently across partitions. This typically 
+involves re-arranging and copying data across executors and machines, making shuffle a complex and 
+costly operation.
+
+#### Background
+
+To understand what happens during the shuffle we can consider the example of the 
+[`reduceByKey`](#ReduceByLink) operation. The `reduceByKey` operation generates a new RDD where all 
+values for a single key are combined into a tuple - the key and the result of executing a reduce 
+function against all values associated with that key. The challenge is that not all values for a 
+single key necessarily reside on the same partition, or even the same machine, but they must be 
+co-located to present a single array per key.
+
+In Spark, data is generally not distributed across partitions to be in the necessary place for a 
```

Comment:
These first couple sentences are a little redundant with the previous paragraph.